docs/proposals/20210716-async-azure-resource-creation-deletion.md

```diff
@@ -0,0 +1,320 @@
+---
+title: Async Azure Resource Creation and Deletion
+authors:
+    - @CecileRobertMichon
+    - @devigned
+reviewers:
+    - TBD
+creation-date: 2021-07-16
+last-updated: 2021-07-26
+status: implementable
+see-also:
+    - https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/1181
+    - https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1067
+---
+
+# Async Azure Resource Creation and Deletion
+
+## <a name='TableofContents'></a>Table of Contents
+
+<!-- vscode-markdown-toc -->
+* [Table of Contents](#TableofContents)
+* [Summary](#Summary)
+* [Motivation](#Motivation)
+	* [Goals](#Goals)
+	* [Non-Goals / Future Work](#Non-GoalsFutureWork)
+* [Proposal](#Proposal)
+	* [User Stories](#UserStories)
+		* [Story 1 - UX of creating an AzureCluster](#Story1-UXofcreatinganAzureCluster)
+		* [Story 2 - Scaling up a MachineDeployment](#Story2-ScalingupaMachineDeployment)
+		* [Story 3 - Deleting an individual Azure Machine Pool Machine](#Story3-DeletinganindividualAzureMachinePoolMachine)
+	* [Implementation Details/Notes/Constraints](#ImplementationDetailsNotesConstraints)
+	* [Proposed API Changes](#ProposedAPIChanges)
+	* [Proposed Controller Changes](#ProposedControllerChanges)
+		* [Context timeouts](#Contexttimeouts)
+		* [Service Reconcile](#ServiceReconcile)
+		* [Service Delete](#ServiceDelete)
+		* [AzureCluster Reconcile](#AzureClusterReconcile)
+	* [Proposed New Conditions](#ProposedNewConditions)
+	* [Open Questions](#OpenQuestions)
+		* [1. What should the timeout durations be?](#Whatshouldthetimeoutdurationsbe)
+* [Alternatives](#Alternatives)
+	* [Parallel reconciliation of Azure services](#ParallelreconciliationofAzureservices)
+		* [Pros](#Pros)
+		* [Cons](#Cons)
+		* [Conclusion](#Conclusion)
+* [Additional Details](#AdditionalDetails)
+	* [Test Plan](#TestPlan)
+* [Implementation History](#ImplementationHistory)
+
+<!-- vscode-markdown-toc-config
+	numbering=false
+	autoSave=false
+	/vscode-markdown-toc-config -->
+<!-- /vscode-markdown-toc -->
+
+## <a name='Summary'></a>Summary
+
+CAPZ reconcilers currently call Azure and wait for each operation before proceeding. We should create/update/delete Azure resources asynchronously, especially for operations that take a long time to complete, such as Virtual Machine creation and deletion.
```

Comment:
Maybe this introductory statement ("CAPZ reconcilers currently call...") could be expanded to make it more clear exactly how bad things are. Maybe:

"Currently, when CAPZ reconciles any Azure resource it does so in a bulk, serialized operation, with the practical effect that reconciling many resources at once results takes a long time, as each operation blocks the next operation."

I can see that this isn't easy to formulate. :/

In any event, my 2¢ is that the current sentence understates things a bit.

Reply:
Based on this Goal below:

"Apply the same asynchronous pattern to all resources"

I am inferring that the above "serial reconciliation" thing is not 100% consistent, and that some resources *do* reconcile discretely and asynchronously. If that's true, then we should probably also include that in the summary as a secondary foundational problem. Like, the fact that we don't implement reconciliation according to a common, consistent strategy/pattern is itself a problem (separate from the problem of picking the best strategy/pattern).

Reply:
This proposal does _not_ aim to reconcile resources in a non-serial (parallel) way. See the Alternatives section where I went into why. Each operation would _still_ block the next operation. This proposal aims to change the fact that a single controller loop right can take up to 90 minutes.

Reply:
> I am inferring that the above "serial reconciliation" thing is not 100% consistent, and that some resources do reconcile discretely and asynchronously.

That's not true. In AzureCluster/AzureMachine controllers, all resources are currently reconciling resources synchronously and serially. This goal highlights that we want to change all of them.

Reply:
Thanks for the clarification!